all-sdk/autotuning.md

```diff
@@ -0,0 +1,431 @@
+  /// Called when deciding what workflow should be evicted from the cache. May return the run id of a workflow to evict.
+  /// If None, or an invalid run id is returned, the default LRU eviction / strategy is used. The id returned must
+  /// be one of the workflows provided in `evictable` (the ones for which there is no actively processing WFT).
+  fn eviction_hint(&self, evictable: &[WorkflowSlotInfo]) -> Option<&str>;
+  /// Called when a workflow is evicted from the cache
+  fn evicted_workflow(&self, evicted: &WorkflowSlotInfo);
+  /// If there is a fixed maximum number of cached workflows, return it.
+  fn max(&self) -> Option<usize>;
+}
+
+/// The user would be allowed to provide either just a slot supplier (using the standard fixed-size cache),
+/// or one of these to handle both themselves.
+/// 
+/// I'm a bit at a loss for a good name
+trait WorkflowSlotAndCacheManager: WorkflowTaskSlotSupplier + WorkflowCacheSizer {}
+```
+
+## Flow
+
+Some context on the current flow (at least in Core). The cache only applies to workflow tasks. This flow should remain
+the same after these changes. The flow makes it clear that a misbehaving `WorkflowSlotAndCacheManager` which allows lots
+of slots but keeps saying the cache should not expand could result in buffering tasks and generally gumming things
+up.
+
+```mermaid
+sequenceDiagram
+  participant T as Temporal
+  participant W as Worker
+  participant S as Slots
+  participant C as CacheSizer
+  W->>S: reserve_slot
+  S-->>W: Allowed
+  W->>T: Poll
+  T-->>W: Response
+  
+  alt Workflow is not already in cache
+    W->>C: `can_allow_workflow` ?
+    C-->>W: I'm full
+    W->>W: Buffer task
+    W->>W: Wait for some workflow to be idle (maybe there is already)
+    W->>C: `eviction_hint`
+    C-->>W: Evict This guy
+    W->>W: Evict workflow
+    W->>C: `evicted_workflow`
+    W->>W: Cache new workflow
+  end
+  
+  W->>S: `mark_slot_used`
+  W->>W: Process workflow task
+  W->>S: `release_slot`
+```
+
+## Default implementations
+
+We should provide a few default implementations:
+
+* Semaphore based SlotSupplier: This is the current implementation. There's a fixed number of slots, that's it.
+* Resource based SlotSupplier: This implementation can look at how close the worker is to some resource limit
+  (% of CPU and/or memory used), and hand out slots if it's still below that limit. This only makes sense for
+  activity slot suppliers, since workflows use little CPU, and the memory they use is the concern of cache
+  management, not the slot supplier.
+* Memory based WorkflowCacheSizer: Like above, but for the sizing of the workflow cache.
+
+## Metrics
+
+Users can add their own metrics behind their implementations now, which is great - but we can also provide some out of
+the box. We can always provide `slots_in_use` by type. Right now we emit the inverse of this,
+`worker_task_slots_available` - we can keep emitting that with a bundled implementation that replicates the old
```

Comment:
> we can keep emitting that with a bundled implementation

A little point: I vote we don't even emit this if they change from the default fixed impl (i.e. the emitting of this should happen in the fixed-slot impl, not part of the general "system" anymore)

Reply:
Why not? It's a nice thing to have automatically get a consistent name if you provide it, same argument as with total cache size. I'm gonna close this comment and the one on the API in favor of the discussion about max cache size since that's got more comments in it and they're the same discussion really.